### PR TITLE
ADM-879 [frontend] Generate data reports for multiple time ranges

### PR DESCRIPTION
**Pull request by Leiqiuhong**
## Summary

Adm-879 [frontend] Generate data reports for multiple time ranges

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

**Review by PengxiWPix — APPROVED** [review]
LGTM

**Review by weiraneve — APPROVED** [review]

**Review by TingyuDong — APPROVED** [review]
LGTM!

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 1]
```diff
@@ -1,118 +1,324 @@
...
+import { useAppSelector } from '@src/hooks/index';
 import { useRef, useState } from 'react';
+import get from 'lodash/get';
 
-export interface useGenerateReportEffectInterface {
+export type PromiseSettledResultWithId<T> = PromiseSettledResult<T> & {
+  id: string;
+};
+
+export interface IUseGenerateReportEffectInterface {
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:18]: rename IUseGenerateReportEffectInterface to IUseGenerateReportEffect

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 2]
```diff
@@ -1,118 +1,324 @@
...
-  const [generalError4Board, setGeneralError4Board] = useState(DEFAULT_MESSAGE);
-  const [generalError4Dora, setGeneralError4Dora] = useState(DEFAULT_MESSAGE);
-  const [generalError4Report, setGeneralError4Report] = useState(DEFAULT_MESSAGE);
-  const [reportData, setReportData] = useState<ReportResponseDTO | undefined>();
+  const configData = useAppSelector(selectConfig);
   const timerIdRef = useRef<number>();
-  let hasPollingStarted = false;
-
-  const startToRequestData = (params: ReportRequestDTO) => {
+  const dateRanges: DateRange = get(configData, 'basic.dateRange', []);
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:80]: could u help to rename `dateRange` to `dateRangeList`

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 3]
```diff
@@ -1,118 +1,324 @@
...
+        });
+      });
     } else {
-      setTimeout4Dora(DEFAULT_MESSAGE);
+      setReportInfos((preReportInfos) => {
+        return preReportInfos.map((reportInfo) => {
+          reportInfo.timeout4Dora = { message: DEFAULT_MESSAGE, shouldShow: true };
+          return reportInfo;
+        });
+      });
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:193]: refactor this part, got `timeout4...` for each case, then set value for it in the `setReportInfos`

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 4]
```diff
@@ -1,118 +1,324 @@
...
 
-  const handleTimeoutError = {
-    [METRIC_TYPES.BOARD]: setTimeout4Board,
-    [METRIC_TYPES.DORA]: setTimeout4Dora,
-    [METRIC_TYPES.ALL]: setTimeout4Report,
+  const updateErrorAfterFetchReport = (
+    res: PromiseSettledResult<ReportCallbackResponse>[],
+    metricTypes: METRIC_TYPES[],
+  ) => {
+    if (res.filter(({ status }) => status === 'rejected').length === 0) return;
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:201]: `'rejected'` as a constant

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 5]
```diff
@@ -1,118 +1,324 @@
...
-    [METRIC_TYPES.DORA]: setGeneralError4Dora,
-    [METRIC_TYPES.ALL]: setGeneralError4Report,
+  const assemblePollingParams = (res: PromiseSettledResult<ReportCallbackResponse>[]) => {
+    const resWithIds: PromiseSettledResultWithId<ReportCallbackResponse>[] = res.map((item, index) => ({
+      ...item,
+      id: reportInfos[index].id,
+    }));
+
+    const fulfilledResponses: PromiseSettledResultWithId<ReportCallbackResponse>[] = resWithIds.filter(
+      ({ status }) => status === 'fulfilled',
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:223]: `'fulfilled'`same as `rejected`

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 6]
```diff
@@ -1,118 +1,324 @@
...
-        } else {
-          timerIdRef.current = window.setTimeout(() => pollingReport(url, interval), interval * 1000);
+  const getNextPollingInfos = (
+    pollingResponsesWithId: PromiseSettledResultWithId<IPollingRes>[],
+    pollingInfos: Record<string, string>[],
+  ) => {
+    const nextPollingInfos: Record<string, string>[] = pollingResponsesWithId
+      .filter(
+        (pollingResponseWithId) =>
+          pollingResponseWithId.status === 'fulfilled' &&
```
Comment (guzhongren) [inline frontend/src/hooks/useGenerateReportEffect.ts:254]: same issues...

**Review by guzhongren — COMMENTED** [review]
Change with the comments

**Comment by guzhongren** [comment]
fix sonar issues

File: frontend/__tests__/containers/ReportStep/ReportStep.test.tsx [hunk 7]
```diff
@@ -403,8 +442,8 @@ describe('Report Step', () => {
       expect(result.current.fetchExportData).toBeCalledWith({
         csvTimeStamp: 0,
         dataType: 'metric',
-        endDate: '',
-        startDate: '',
+        endDate: '2024-02-28T23:59:59.999+08:00',
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/containers/ReportStep/ReportStep.test.tsx:445]: Where exactly is it repeated?

**Review by yulongcai — APPROVED** [review]

File: frontend/__tests__/containers/ReportStep/ReportStep.test.tsx [hunk 8]
```diff
@@ -559,14 +611,78 @@ describe('Report Step', () => {
...
     });
+
+    it('should not show notification in sending request', async () => {
+      reportHook.current.hasPollingStarted = true;
+      setup(REQUIRED_DATA_LIST, [emptyValueDateRange]);
+
+      expect(addNotification).toHaveBeenCalledTimes(0);
+    });
+
+    it('should not show notification when the requests all failed', () => {
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/containers/ReportStep/ReportStep.test.tsx]: done

File: frontend/__tests__/containers/ReportStep/ReportStep.test.tsx [hunk 9]
```diff
@@ -559,14 +611,78 @@ describe('Report Step', () => {
...
+      reportHook.current.hasPollingStarted = false;
+      reportHook.current.reportInfos[0].reportData = undefined;
+      reportHook.current.reportInfos[1].reportData = undefined;
+      setup(REQUIRED_DATA_LIST, [fullValueDateRange]);
+      expect(addNotification).toHaveBeenCalledTimes(0);
+    });
+
+    it('should show "file will expire ..." notification when request is successful', () => {
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/containers/ReportStep/ReportStep.test.tsx]: done

File: frontend/src/hooks/useGenerateReportEffect.ts (hunk 3 above)
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/src/hooks/useGenerateReportEffect.ts (hunk 4 above)
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/src/hooks/useGenerateReportEffect.ts (hunk 5 above)
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/src/hooks/useGenerateReportEffect.ts (hunk 6 above)
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/__tests__/containers/MetricsStepper/MetricsStepper.test.tsx [hunk 10]
```diff
@@ -56,6 +57,11 @@ const mockValidationCheckContext = {
   getDuplicatedPipeLineIds: jest.fn().mockReturnValue([]),
 };
 
+const mockDateRange = {
+  startDate: dayjs().startOf('date').add(0, 'day').format(COMMON_TIME_FORMAT),
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/containers/MetricsStepper/MetricsStepper.test.tsx]: done

File: frontend/__tests__/containers/ReportStep/ReportStep.test.tsx [hunk 11]
```diff
@@ -559,14 +611,78 @@ describe('Report Step', () => {
...
+      reportHook.current.hasPollingStarted = false;
+      setup(REQUIRED_DATA_LIST, [fullValueDateRange]);
+      expect(addNotification).toHaveBeenCalledWith({
+        message: MESSAGE.EXPIRE_INFORMATION(30),
+      });
+    });
+
+    it('should not show notifications after shown once', () => {
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/containers/ReportStep/ReportStep.test.tsx]: done

File: frontend/__tests__/hooks/useGenerateReportEffect.test.tsx [hunk 12]
```diff
@@ -18,212 +30,250 @@ const MOCK_GENERATE_REPORT_REQUEST_PARAMS_WITH_DORA_METRIC_TYPE = {
...
+    wrapper: Wrapper,
+  });
+
+const dateRanges = [
+  {
+    startDate: '2024-02-04T00:00:00.000+08:00',
+    endDate: '2024-02-17T23:59:59.999+08:00',
+  },
+  {
+    startDate: '2024-02-18T00:00:00.000+08:00',
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/__tests__/hooks/useGenerateReportEffect.test.tsx]: done

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 13]
```diff
@@ -1,118 +1,324 @@
...
   startToRequestData: (params: ReportRequestDTO) => void;
   stopPollingReports: () => void;
-  timeout4Board: string;
-  timeout4Dora: string;
-  timeout4Report: string;
-  generalError4Board: string;
-  generalError4Dora: string;
-  generalError4Report: string;
+  reportInfos: IReportInfo[];
+  shutReportInfosErrorStatus: (id: string, errorKey: string) => void;
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/src/hooks/useGenerateReportEffect.ts (hunk 1 above)
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

File: frontend/src/hooks/useGenerateReportEffect.ts [hunk 14]
```diff
@@ -1,118 +1,324 @@
...
+  shutBoardMetricsError: (id: string) => void;
+  shutPipelineMetricsError: (id: string) => void;
+  shutSourceControlMetricsError: (id: string) => void;
+  hasPollingStarted: boolean;
+}
+
+export interface IReportError {
+  timeout4Board: { message: string; shouldShow: boolean };
```
[the remark this answers is not in the document]
Comment (Leiqiuhong) [reply frontend/src/hooks/useGenerateReportEffect.ts]: done

**Pull request merged**